src/libstore/build.cc

```diff
@@ -3004,6 +3004,29 @@ PathSet parseReferenceSpecifiers(Store & store, const BasicDerivation & drv, con
 }
 
 
+static void moveCheckToStore(const Path & src, const Path & dst)
+{
+    /* For the rename of directory to succeed, we must be running as root or
+       the directory must be made temporarily writable (to update the
+       directory's parent link ".."). */
+    struct stat st;
+    if (lstat(src.c_str(), &st) == -1) {
+        throw SysError(format("getting attributes of path '%1%'") % src);
+    }
+
+    bool changePerm = (geteuid() && S_ISDIR(st.st_mode) && !(st.st_mode & S_IWUSR));
+
+    if (changePerm)
+        chmod_(src, st.st_mode | S_IWUSR);
+
+    if (rename(src.c_str(), dst.c_str()))
+        throw SysError(format("renaming '%1%' to '%2%'") % src % dst);
+
+    if (changePerm)
```

Comment:
I'm not sure if it's important to restore permissions here, the check path is for debugging only and isn't registered. Outputs of failed builds, etc. that have not been garbage collected yet generally also leave writable paths behind.

Reply:
I appreciate your comment. I wasn't sure it was necessary to restore permissions but figured it was conservative and couldn't hurt. I'll probably leave it as is for now since it has been tested and seems slightly cleaner.

Reply:
Yeah, doesn't hurt. Don't forget to re-enable that test after rebasing. 😄

Reply:
Thanks for all of your help. Will do.